ngraph_bridge/ngraph_encapsulate_impl.cc

```diff
@@ -0,0 +1,433 @@
+#include "ngraph_bridge/ngraph_encapsulate_impl.h"
+#include "ngraph_bridge/ngraph_encapsulate_op.h"
+#include "ngraph_bridge/ngraph_mark_for_clustering.h"
+#include "ngraph_bridge/ngraph_timer.h"
+#include "ngraph_bridge/ngraph_utils.h"
+
+#if defined(NGRAPH_TF_ENABLE_VARIABLES_AND_OPTIMIZERS)
+#include "ngraph_bridge/enable_variable_ops/ngraph_catalog.h"
+#include "ngraph_bridge/enable_variable_ops/ngraph_var.h"
+#endif
+
+using namespace std;
+namespace ng = ngraph;
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+// Ngraph Encapsulate Implementation class for EncapsulateOp class
+//---------------------------------------------------------------------------
+//  NGraphEncapsulateImpl::ctor
+//---------------------------------------------------------------------------
+NGraphEncapsulateImpl::NGraphEncapsulateImpl(string name)
+    : m_graph(OpRegistry::Global()),
+      m_freshness_tracker(nullptr),
+      m_name(name) {
+  my_instance_id = s_instance_count;
+  s_instance_count++;
+}
+
+// Get tensorflow input tensors, input shapes, static_inputs to Compute
+// Signature
+Status NGraphEncapsulateImpl::ComputeSignature(
+    const std::vector<Tensor>& tf_input_tensors,
+    std::vector<TensorShape>& input_shapes,
+    std::vector<const Tensor*>& static_input_map,
+    std::stringstream& signature_ss) {
+  // Get the inputs
+  for (int i = 0; i < tf_input_tensors.size(); i++) {
+    const Tensor& input_tensor = tf_input_tensors[i];
+    input_shapes.push_back(input_tensor.shape());
+    for (const auto& x : input_tensor.shape()) {
+      signature_ss << x.size << ",";
+    }
+    signature_ss << ";";
+  }
+
+  signature_ss << "/";
+
+  static_input_map.resize(tf_input_tensors.size());
+  for (int i = 0; i < tf_input_tensors.size(); i++) {
+    const Tensor& input_tensor = tf_input_tensors[i];
+    if (get_static()[i]) {
+      static_input_map[i] = &input_tensor;
+      TF_RETURN_IF_ERROR(TensorToStream(signature_ss, input_tensor));
+      signature_ss << ";";
+    }
+  }
+  return Status::OK();
+}
+
+// Calls Compute Signature and gets ngraph executable
+Status NGraphEncapsulateImpl::GetNgExecutable(
+    const std::vector<Tensor>& tf_input_tensors,
+    std::vector<TensorShape>& input_shapes,
+    std::vector<const Tensor*>& static_input_map,
+    ng::runtime::Backend*& op_backend,
+    std::shared_ptr<ngraph::runtime::Executable>& ng_exec) {
+  std::stringstream signature_ss;
+  string signature;
+
+  std::shared_ptr<ngraph::Function> ng_function;
+  std::shared_ptr<ngraph::runtime::Executable> evicted_ng_exec;
+
+  NGRAPH_VLOG(4) << "GetNgExecutable: Got backend of type: "
+                 << get_op_backend_name();
+  op_backend = BackendManager::GetBackend(get_op_backend_name());
+
+  // Compute Signature
+  TF_RETURN_IF_ERROR(ComputeSignature(tf_input_tensors, input_shapes,
+                                      static_input_map, signature_ss));
+  signature = signature_ss.str();
+
+  if (NGRAPH_VLOG_IS_ON(5)) {
```

Comment:
I think the way VLOG is implemented, the output string will be created regardless. Then it will be displayed if the log is on. Since the signature printing can be significantly expensive, it's good to have this check. @sindhu-nervana can you please verify?